File: src/no-more-bets/utils/match_matcher.py
"""Utility class for matching teams and data across different sources."""
import logging
from typing import Dict, List, Optional, Mapping
from rapidfuzz import process
from models.rotowire import GameLineup
from models.soccerdata import UpcomingMatchPreview, LeagueMatchPreviews
from models.fotmob import Club

logger = logging.getLogger(__name__)


class MatchMatcher:
    """Utility class for matching teams and data across different sources."""
    
    @staticmethod
    def teams_key(home: str, away: str) -> frozenset[str]:
        """Create a normalized key for team matching.
        
        Parameters
        ----------
        home : str
            Home team name.
        away : str
            Away team name.
            
        Returns
        -------
        frozenset[str]
            Normalized key for team matching.
        """
        return frozenset({home.lower().strip(), away.lower().strip()})
    
    @staticmethod
    def build_lineup_index(lineups: List[GameLineup]) -> Dict[frozenset[str], GameLineup]:
        """Build an index of lineups by team names.
        
        Parameters
        ----------
        lineups : List[GameLineup]
            List of game lineups to index.
            
        Returns
        -------
        Dict[frozenset[str], GameLineup]
            Dictionary mapping team keys to lineups.
        """
        return {
            MatchMatcher.teams_key(lineup.home_team.team_name, lineup.away_team.team_name): lineup
            for lineup in lineups
        }
    
    @staticmethod
    def find_lineup(
        match_preview: UpcomingMatchPreview,
        lineup_index: Mapping[frozenset[str], GameLineup]
    ) -> Optional[GameLineup]:
        """Find matching lineup for a match preview.
        
        Parameters
        ----------
        match_preview : UpcomingMatchPreview
            Match preview to find lineup for.
        lineup_index : Mapping[frozenset[str], GameLineup]
            Index of lineups by team keys.
            
        Returns
        -------
        Optional[GameLineup]
            Matching lineup if found, None otherwise.
        """
        key = MatchMatcher.teams_key(
            match_preview.teams.home.name,
            match_preview.teams.away.name
        )
        
        if key in lineup_index:
            return lineup_index[key]
        
        # Fuzzy matching
        search_str = " vs ".join(sorted(key))
        candidates = {" vs ".join(sorted(k)): lineup_index[k] for k in lineup_index.keys()}
        
        result = process.extractOne(search_str, candidates.keys(), score_cutoff=85)
        if result:
            return candidates[result[0]]
        
        return None
    
    @staticmethod
    def find_soccerdata_match(
        home_team_name: str,
        away_team_name: str,
        upcoming_league_matches: List[LeagueMatchPreviews]
    ) -> Optional[UpcomingMatchPreview]:
        """Find matching match from upcoming league matches by team names.
        
        Parameters
        ----------
        home_team_name : str
            Home team name.
        away_team_name : str
            Away team name.
        upcoming_league_matches : List[LeagueMatchPreviews]
            List of league match previews to search.
            
        Returns
        -------
        Optional[UpcomingMatchPreview]
            Matching match preview if found, None otherwise.
        """
        key = MatchMatcher.teams_key(home_team_name, away_team_name)
        
        # Search through all leagues and matches
        for league in upcoming_league_matches:
            for match in league.match_previews:
                match_key = MatchMatcher.teams_key(match.teams.home.name, match.teams.away.name)
                if match_key == key:
                    return match
        
        # Fuzzy matching
        search_str = " vs ".join(sorted(key))
        candidates = {}
        for league in upcoming_league_matches:
            for match in league.match_previews:
                match_key = MatchMatcher.teams_key(match.teams.home.name, match.teams.away.name)
                candidate_str = " vs ".join(sorted(match_key))
                candidates[candidate_str] = match
        
        result = process.extractOne(search_str, candidates.keys(), score_cutoff=85)
        if result:
            return candidates[result[0]]
        
        return None
    
    @staticmethod
    def find_fotmob_club(team_name: str, fotmob_clubs: List[Club]) -> Optional[Club]:
        """Find matching club from FotMob clubs by team name using fuzzy matching.
        
        Parameters
        ----------
        team_name : str
            Team name to search for.
        fotmob_clubs : List[Club]
            List of FotMob clubs to search.
            
        Returns
        -------
        Optional[Club]
            Matching club if found, None otherwise.
        """
        # Normalize team name for better matching
        normalized_search = team_name.lower().strip()
        
        # Exact match first (case-insensitive)
        for club in fotmob_clubs:
            if club.team_name.lower().strip() == normalized_search:
                return club
        
        # Try partial match (if search name is contained in club name or vice versa)
        for club in fotmob_clubs:
            normalized_club = club.team_name.lower().strip()
            if normalized_search in normalized_club or normalized_club in normalized_search:
                return club
        
        # Fuzzy matching with lower threshold for better matching
        candidates = {club.team_name: club for club in fotmob_clubs}
        result = process.extractOne(team_name, candidates.keys(), score_cutoff=70)
        if result:
            logger.debug(f"Fuzzy matched '{team_name}' to '{result[0]}' (score: {result[1]:.1f})")
            return candidates[result[0]]
        
        return None

File: src/NoMoreBets/Features/MatchAnalysis/MatchMatcher/MatchMatcher.cs
using FuzzySharp;
using NoMoreBets.Domain.Clubs;
using NoMoreBets.Features.Fotmob.GetFotmobLeagueTable.Dtos;
using NoMoreBets.Features.Fotmob.GetFotmobXgStats.Dtos;
using NoMoreBets.Features.Rotowire.Model;
using NoMoreBets.Features.SoccerData.Model;

namespace NoMoreBets.Features.MatchAnalysis.MatchMatcher;

/// <summary>
/// Matches teams and data across sources using exact key match first, then FuzzySharp.
/// </summary>
public sealed class MatchMatcher : IMatchMatcher
{
  private const int LineupAndSoccerDataScoreCutoff = 85;
  private const int FotmobScoreCutoff = 70;
  private readonly ILogger<MatchMatcher> _logger;

  public MatchMatcher(ILogger<MatchMatcher> logger)
  {
    _logger = logger;
  }

  /// <inheritdoc />
  public IReadOnlyDictionary<TeamKey, GameLineup> BuildLineupIndex(IReadOnlyList<GameLineup> lineups)
  {
    var dict = new Dictionary<TeamKey, GameLineup>(lineups.Count);
    foreach (var lineup in lineups)
    {
      var key = new TeamKey(lineup.HomeTeamName, lineup.AwayTeamName);
      dict[key] = lineup;
    }
    return dict;
  }

  /// <inheritdoc />
  public GameLineup? FindLineup(string home, string away, IReadOnlyDictionary<TeamKey, GameLineup> index)
  {
    var key = new TeamKey(home, away);
    if (index.TryGetValue(key, out var lineup))
    {
      return lineup;
    }

    var searchStr = key.ToSearchString();
    var candidates = index.ToDictionary(
        kv => kv.Key.ToSearchString(),
        kv => kv.Value,
        StringComparer.Ordinal);

    if (candidates.Count == 0)
    {
      _logger.LogError("No lineup candidates were found for {Home} vs {Away}", home, away);
      return null;
    }

    var keys = candidates.Keys.ToList();
    var best = Process.ExtractOne(searchStr, keys, s => s, cutoff: LineupAndSoccerDataScoreCutoff);
    if (best is null)
    {
      _logger.LogError("No matching lineup found for {Home} vs {Away}", home, away);
      return null;
    }

    var value = best.Value;
    if (value != null && best.Score >= LineupAndSoccerDataScoreCutoff && candidates.TryGetValue(value, out var found))
    {
      return found;
    }

    _logger.LogError("No matching lineup found for {Home} vs {Away}", home, away);
    return null;
  }

  /// <inheritdoc />
  public UpcomingMatchPreview? FindSoccerDataMatch(string home, string away, IReadOnlyList<LeagueMatchPreviews> leagues)
  {
    var key = new TeamKey(home, away);
    var searchStr = key.ToSearchString();

    foreach (var league in leagues)
    {
      foreach (var match in league.MatchPreviews)
      {
        var matchKey = new TeamKey(match.Teams.Home.Name, match.Teams.Away.Name);
        if (matchKey.Equals(key))
        {
          return match;
        }
      }
    }

    var candidates = new Dictionary<string, UpcomingMatchPreview>(StringComparer.Ordinal);
    foreach (var league in leagues)
    {
      foreach (var match in league.MatchPreviews)
      {
        var k = new TeamKey(match.Teams.Home.Name, match.Teams.Away.Name);
        candidates[k.ToSearchString()] = match;
      }
    }

    if (candidates.Count == 0)
    {
      _logger.LogError("No candidates were found for {Home} vs {Away}", home, away);
      return null;
    }
    var keys = candidates.Keys.ToList();
    var best = Process.ExtractOne(searchStr, keys, s => s, cutoff: LineupAndSoccerDataScoreCutoff);
    if (best == null)
    {
      _logger.LogError("No matching soccer data match found for {Home} vs {Away}. Bets was null", home, away);
      return null;
    }

    var value = best.Value;
    if (value != null && best.Score >= LineupAndSoccerDataScoreCutoff && candidates.TryGetValue(value, out var found))
    {
      return found;
    }

    _logger.LogError("No matching soccer data match found for {Home} vs {Away}", home, away);
    return null;
  }

  /// <inheritdoc />
  public ClubDto? FindFotmobClub(string teamName, IReadOnlyList<ClubDto> clubs)
  {
    if (clubs.Count == 0)
    {
      _logger.LogError("No clubs to perform search from");
      return null;
    }

    var normalized = (teamName ?? string.Empty).Trim().ToLowerInvariant();
    foreach (var club in clubs)
    {
      if (club.TeamName.Trim().ToLowerInvariant() == normalized)
      {
        return club;
      }
    }

    foreach (var club in clubs)
    {
      var cn = club.TeamName.Trim().ToLowerInvariant();
      if (normalized.Contains(cn, StringComparison.Ordinal) || cn.Contains(normalized, StringComparison.Ordinal))
      {
        return club;
      }
    }

    var choices = clubs.Select(c => c.TeamName).ToArray();
    if (choices.Length == 0)
    {
      _logger.LogError("No club choices selected");
      return null;
    }

    var best = Process.ExtractOne(teamName, choices, s => s ?? "", cutoff: FotmobScoreCutoff);
    if (best == null)
    {
      _logger.LogError("No matching club data found for {Club}", teamName);
      return null;
    }
    var idx = best.Index;
    if (idx >= 0 && idx < clubs.Count && best.Score >= FotmobScoreCutoff)
    {
      return clubs[idx];
    }

    _logger.LogError("No matching club data found for {Club}", teamName);
    return null;
  }

  /// <inheritdoc />
  public Club FindClub(string teamName, IReadOnlyList<Club> clubs)
  {
    var trimmed = (teamName ?? string.Empty).Trim();
    if (clubs.Count == 0)
    {
      throw new InvalidOperationException(
        $"No clubs in league to match. Cannot resolve team '{trimmed}'.");
    }

    foreach (var club in clubs)
    {
      if (string.Equals(club.Name.Trim(), trimmed, StringComparison.OrdinalIgnoreCase))
      {
        return club;
      }
    }

    var names = clubs.Select(c => c.Name).ToArray();
    var best = Process.ExtractOne(trimmed, names, s => s ?? "", cutoff: FotmobScoreCutoff);
    if (best != null && best.Score >= FotmobScoreCutoff && best.Index >= 0 && best.Index < clubs.Count)
    {
      return clubs[best.Index];
    }

    throw new InvalidOperationException(
      $"No matching club found for '{trimmed}' among {clubs.Count} clubs. Ensure the team exists in the league.");
  }

  /// <inheritdoc />
  public XgStatsDto? FindXgStats(string teamName, IReadOnlyList<XgStatsDto> xgStats)
  {
    if (xgStats.Count == 0)
    {
      _logger.LogError("No xG stats to perform search from");
      return null;
    }

    var normalized = (teamName ?? string.Empty).Trim().ToLowerInvariant();
    foreach (var stat in xgStats)
    {
      var tn = (stat.TeamName ?? string.Empty).Trim().ToLowerInvariant();
      if (tn == normalized)
      {
        return stat;
      }
    }

    foreach (var stat in xgStats)
    {
      var tn = (stat.TeamName ?? string.Empty).Trim().ToLowerInvariant();
      if (normalized.Contains(tn, StringComparison.Ordinal) || tn.Contains(normalized, StringComparison.Ordinal))
      {
        return stat;
      }
    }

    var choices = xgStats.Select(s => s.TeamName ?? string.Empty).ToArray();
    if (choices.Length == 0)
    {
      _logger.LogError("No xG stats choices for matching");
      return null;
    }

    var best = Process.ExtractOne(teamName, choices, s => s ?? "", cutoff: FotmobScoreCutoff);
    if (best == null)
    {
      _logger.LogError("No matching xG stats found for {Club}", teamName);
      return null;
    }
    var idx = best.Index;
    if (idx >= 0 && idx < xgStats.Count && best.Score >= FotmobScoreCutoff)
    {
      return xgStats[idx];
    }

    _logger.LogError("No matching xG stats found for {Club}", teamName);
    return null;
  }

  /// <inheritdoc />
  public T? FindBestMatch<T>(string home, string away, IReadOnlyList<(string HomeName, string AwayName, T Value)> candidates)
  {
    if (candidates.Count == 0)
    {
      return default;
    }

    var key = new TeamKey(home, away);
    var searchStr = key.ToSearchString();

    foreach (var c in candidates)
    {
      var matchKey = new TeamKey(c.HomeName, c.AwayName);
      if (matchKey.Equals(key))
      {
        return c.Value;
      }
    }

    var dict = new Dictionary<string, T>(StringComparer.Ordinal);
    foreach (var c in candidates)
    {
      var k = new TeamKey(c.HomeName, c.AwayName);
      dict[k.ToSearchString()] = c.Value;
    }

    var keys = dict.Keys.ToList();
    var best = Process.ExtractOne(searchStr, keys, s => s, cutoff: LineupAndSoccerDataScoreCutoff);
    if (best is null)
    {
      return default;
    }

    var value = best.Value;
    if (value != null && best.Score >= LineupAndSoccerDataScoreCutoff && dict.TryGetValue(value, out var found))
    {
      return found;
    }

    return default;
  }
}
